### PR TITLE
Fixed session init for APIs

### PR DESCRIPTION
**Pull request by edgardmessias**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #3688

With this, the session is initialized only once
First check the session token, for later, initialize the session
For get the session, need parse the content, to run later.

### CODE REVIEW

**Comment by trasher** [comment]
API unit tests are failing

**Comment by edgardmessias** [comment]
I will fix it

**Comment by edgardmessias** [comment]
@trasher, I fixed, but not understand why not work on PHP 7.2

**Comment by trasher** [comment]
Hum... There is a warning; maybe is this related. I'll take a look

**Comment by edgardmessias** [comment]
The problem: `PHP Warning:  session_save_path(): Cannot change save path when session is active`

**Comment by trasher** [comment]
Yes, there are lots of output in the local webserver; this has to be fixed... But on my computer, tests are OK even with those issues.

**Comment by edgardmessias** [comment]
Fixed, wait result

File: inc/api.class.php [hunk 1]
```diff
@@ -257,30 +257,11 @@ protected function retrieveSession() {
          $current = session_id();
          $session = trim($this->parameters['session_token']);
 
-         if (file_exists(GLPI_ROOT . '/inc/downstream.php')) {
```
Comment (trasher) [inline inc/api.class.php]: You can't drop all of this. See http://glpi-install.readthedocs.io/en/latest/install/index.html#files-and-directories-locations

File: tests/DbTestCase.php [hunk 2]
```diff
@@ -84,7 +84,7 @@ public function afterTestMethod($method) {
       // Cleanup log directory
       foreach (glob(GLPI_LOG_DIR . '/*.log') as $file) {
          if (file_exists($file)) {
-            unlink($file);
+            @unlink($file);
```
Comment (trasher) [inline tests/DbTestCase.php]: I do not agree; and this should really not be required.

File: inc/api.class.php (hunk 1 above)
Comment (edgardmessias) [reply inc/api.class.php]: I will revert

**Comment by edgardmessias** [comment]
@trasher, fixed and tests passing

**Comment by DIOHz0r** [comment]
Hi @trasher,

I also did another PR #3697 related with the same issue, Could you check both PR and decide which one will be merged?.

Thank tou.

File: inc/apirest.class.php [hunk 3]
```diff
@@ -55,6 +55,20 @@ public static function getTypeName($nb = 0) {
       return __('Rest API');
    }
 
+   public function prepare() {
+      parent::prepare();
+
+      foreach ($_FILES as $filename => $files) {
```
Comment (orthagh) [inline inc/apirest.class.php]: Is this move of fileupload var parsing is needed ?

Comment (edgardmessias) [reply inc/apirest.class.php]: Yes, because the`GLPIUploadHandler::uploadFiles` call document validation, and only validade after load config

Comment (orthagh) [reply inc/apirest.class.php]: ok, just a suggestion so, maybe we could move this code to a separate function `manageUploadedFiles` (for example) and call after the initialization.

File: inc/api.class.php [hunk 4]
```diff
@@ -84,9 +85,18 @@
     *
     * @return void
     */
-   public function __construct() {
+   public function prepare() {
```
Comment (orthagh) [inline inc/api.class.php]: pretty subjective, i dont like `prepare` term for the behavior.
I suggest `initApi`

Comment (edgardmessias) [reply inc/api.class.php]: I will make this.

**Comment by edgardmessias** [comment]
@orthagh, done.

File: inc/apirest.class.php [hunk 5]
```diff
@@ -55,6 +55,23 @@ public static function getTypeName($nb = 0) {
       return __('Rest API');
    }
 
+   /**
+    * Upload and validate files from request and append to $this->parameters['input']
+    *
+    * @return void
+    */
+   public function manageUploadedFiles() {
```
Comment (orthagh) [inline inc/apirest.class.php]: last thing (sorry)
make this function abstract in api.class, and do it also in xmlrpc (empty of course)

**Comment by orthagh** [comment]
great, i'll do some additional tests but we'll could merge.

It could be done in another pr, but ideally, we could make a test for #3688 issue.
I can take care of that task.

**Comment by edgardmessias** [comment]
@orthagh, done

**Pull request merged**